### DISCUSSION TITLE
Any Chance This Could Work on Pinephone?

### DISCUSSION THREAD

**Discussion by ChristianSilvermoon** [Wohlstand/TheXTech#35] [General]
I've tried compiling this in [Termux](https://github.com/termux/termux-app) on Android before... and now on [Mobian](https://mobian-project.org/) on my [Pinephone](https://www.pine64.org/pinephone/)... but I just can't seem to get the game to compile on ARM at all.

I tried to compile tag v1.3.3.1, and... well, no luck.

![20210205_21h11m12s_grim](https://user-images.githubusercontent.com/30162808/107109185-2b8c4f80-67fb-11eb-8077-31d96fa9f383.png)
![20210205_21h12m16s_grim](https://user-images.githubusercontent.com/30162808/107109188-31823080-67fb-11eb-88d0-aa8474f916ec.png)
![20210205_21h41m53s_grim](https://user-images.githubusercontent.com/30162808/107109189-36df7b00-67fb-11eb-9aec-33183d63e037.png)

I logged the compilation with `ninja | tee ~/comp.log` so I could share it.
[comp.log](https://github.com/Wohlstand/TheXTech/files/5936264/comp.log)

I'm... not really the best at stuff, so if someone could help me understand, that'd be really great ^^

**Comment by Wohlstand** [comment]
Looks like OPUS fails because of Assembly files:

![SmartSelect_20210206-123523_ES File Explorer Pro.jpg](https://user-images.githubusercontent.com/6751442/107114603-1e617600-6878-11eb-918b-7a9970549015.jpg)

They getting normally built on Linux (I built the game on Orange Pi PC from under Ubuntu 16.04 on ARM) and for Android with the Gradle and NDK. Maybe I need to finish my work on the ARM detector for the OPUS? It was incomplete and partial...

**Reply by ChristianSilvermoon** [reply]
Any other information I can give that might help I'm happy to do my best to provide ^^

EDIT: Would it be possible to use this in some way (from the Debian Bullseye repo)?
```
libopus-dev/testing 1.3.1-0.1 arm64
```
?

**Reply by Wohlstand** [reply]
You can install the rest of AudioCodecs manually from the system, however, the libADLMIDI, libOPNMIDI, libGME, and also libXMP, you need to build separately. The build at me is intended to compile the whole stuff through the one build, but for some special cases, here is an ability to re-use libraries from the system.

So:
* Install the `libogg-dev libvorbis-dev libflac-dev libopus-dev libopusfile-dev libmad0-dev libmpg123-dev libsdl2-dev libpng-dev libjpeg-dev libmodplug-dev libmikmod-dev`
* Build and install the [libADLMIDI](https://github.com/Wohlstand/libADLMIDI) and [libOPNMIDI](https://github.com/Wohlstand/libOPNMIDI) from the sources (using a common CMake build)
* Build and install the [libXMP](https://github.com/libxmp/libxmp) from the sources, using an Autotools way
* Build and install the [libGME](https://bitbucket.org/mpyne/game-music-emu/wiki/Home) from the sources using CMake way
* Build and install the [MixerX](https://github.com/WohlSoft/SDL-Mixer-X) library by using the common CMake build, all necessary libraries should be recognized from the system
* Build and install the [FreeImageLite](https://github.com/WohlSoft/libFreeImage) by the regular CMake build
* While building the game, use the `-DUSE_SYSTEM_LIBS=ON -DUSE_SYSTEM_SDL2=ON` CMake flags

**Reply by ChristianSilvermoon** [reply]
Oh boy... I'll give it a shot, I'm not really great at compiling stuff, but I guess I gotta learn somehow XD

I've gotten stuck at libXMP... I've tried `autoconf` and `./configure` and it appears to not be able to guess my system type, I suppose?

```
checking build system type... ./config.guess: unable to guess system type

This script, last modified 2009-11-20, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD
and
  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD

If the version you run (./config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2009-11-20

uname -m = aarch64
uname -r = 5.10-sunxi64
uname -s = Linux
uname -v = #2 SMP PREEMPT Tue Jan 12 09:55:56 UTC 2021

/usr/bin/uname -p = unknown
/bin/uname -X     = 

hostinfo               = 
/bin/universe          = 
/usr/bin/arch -k       = 
/bin/arch              = aarch64
/usr/bin/oslevel       = 
/usr/convex/getsysinfo = 

UNAME_MACHINE = aarch64
UNAME_RELEASE = 5.10-sunxi64
UNAME_SYSTEM  = Linux
UNAME_VERSION = #2 SMP PREEMPT Tue Jan 12 09:55:56 UTC 2021
configure: error: cannot guess build type; you must specify one
```

I am unsure of what to do here

EDIT: `./configure --build=arm-pc-linux-gnu` seems to have gotten me somewhere...

**Reply by Wohlstand** [reply]
For the worst case you can ignore the libXMP build or try to install an old version from the package manager `libxmp-dev`. Keep a note its buggy in some places. So, the latest mainstream build is suggested.

**Reply by Wohlstand** [reply]
The ModPlug and MikMod will be used then if you'll don't build the libXMP.

**Reply by ChristianSilvermoon** [reply]
It actually built with that
```
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/  ..

-- The C compiler identification is GNU 10.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
== Using default build configuration which is a Release!
AudioCodecs will NOT be used. Libraries will be recognized in the system.
-- Checking for module 'sdl2'
--   Found sdl2, version 2.0.14
== SDL2: /usr/include/SDL2 SDL2 ==
-- Looking for ov_open in vorbisfile
-- Looking for ov_open in vorbisfile - found
Vorbis: /usr/include /usr/lib/aarch64-linux-gnu/libvorbis.so
== using Vorbis ==
-- Looking for op_open_file in opusfile
-- Looking for op_open_file in opusfile - found
Opus: /usr/include /usr/lib/libopusfile.so /usr/lib/libopusfile.so
== using Opus ==
-- Looking for FLAC__format_sample_rate_is_valid in FLAC
-- Looking for FLAC__format_sample_rate_is_valid in FLAC - found
FLAC: /usr/include /usr/lib/aarch64-linux-gnu/libFLAC.so
== using FLAC ==
-- Looking for mad_frame_init in mad
-- Looking for mad_frame_init in mad - found
MAD: /usr/include /usr/lib/aarch64-linux-gnu/libmad.so
== using MAD ==
-- Looking for ModPlug_Load in modplug
-- Looking for ModPlug_Load in modplug - found
ModPlug: /usr/include /usr/lib/aarch64-linux-gnu/libmodplug.so
== using libModPlug ==
-- Looking for gme_load_file in gme
-- Looking for gme_load_file in gme - found
GME: LIBGME_INCLUDE_DIR-NOTFOUND /usr/local/lib/libgme.so
== using GME ==
== using CMD Music ==
-- Looking for fork
-- Looking for fork - found
-- Looking for adl_init in ADLMIDI
-- Looking for adl_init in ADLMIDI - not found
ADLMIDI: /usr/local/include /usr/local/lib/libADLMIDI.a
-- Looking for opn_init in OPNMIDI
-- Looking for opn_init in OPNMIDI - not found
OPNMIDI: /usr/local/include /usr/local/lib/libOPNMIDI.a
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mobian/Dev/github.com/Wohlstand/SDL-Mixer-X/build
```

This feels like it suggests a file is missing, but that file exists.
`GME: LIBGME_INCLUDE_DIR-NOTFOUND /usr/local/lib/libgme.so`

`make -j4` fails with
```
make -j4
Scanning dependencies of target SDL2_mixer_ext
Scanning dependencies of target SDL2_mixer_ext_Static
[ 11%] Building C object CMakeFiles/SDL2_mixer_ext.dir/src/codecs/load_voc.c.o
[ 11%] Building C object CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/load_aiff.c.o
[ 11%] Building C object CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/load_voc.c.o
[ 11%] Building C object CMakeFiles/SDL2_mixer_ext.dir/src/codecs/load_aiff.c.o
[ 13%] Building C object CMakeFiles/SDL2_mixer_ext.dir/src/codecs/music_cmd.c.o
[ 16%] Building C object CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/music_cmd.c.o
[ 19%] Building C object CMakeFiles/SDL2_mixer_ext.dir/src/codecs/music_flac.c.o
[ 22%] Building C object CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/music_flac.c.o
[ 25%] Building C object CMakeFiles/SDL2_mixer_ext.dir/src/codecs/music_gme.c.o
[ 27%] Building C object CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/music_gme.c.o
/home/mobian/Dev/github.com/Wohlstand/SDL-Mixer-X/src/codecs/music_gme.c:33:10: fatal error: gme.h: No such file or directory
   33 | #include <gme.h>
      |          ^~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/SDL2_mixer_ext.dir/build.make:134: CMakeFiles/SDL2_mixer_ext.dir/src/codecs/music_gme.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/mobian/Dev/github.com/Wohlstand/SDL-Mixer-X/src/codecs/music_gme.c:33:10: fatal error: gme.h: No such file or directory
   33 | #include <gme.h>
      |          ^~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/SDL2_mixer_ext_Static.dir/build.make:134: CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/music_gme.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 30%] Building C object CMakeFiles/SDL2_mixer_ext_Static.dir/src/codecs/music_id3tag.c.o
make[1]: *** [CMakeFiles/Makefile2:97: CMakeFiles/SDL2_mixer_ext.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:124: CMakeFiles/SDL2_mixer_ext_Static.dir/all] Error 2
make: *** [Makefile:149: all] Error 2
```

**Reply by Wohlstand** [reply]
Did you built and installed the libGME?

**Reply by ChristianSilvermoon** [reply]
Eeyup,

and
```
file /usr/local/lib/libgme.so
/usr/local/lib/libgme.so: symbolic link to libgme.so.0
```
it's definitely there, hmm...

**Reply by Wohlstand** [reply]
What about the public header? `gme.h`, is it at /usr/local/include/ ?

**Reply by ChristianSilvermoon** [reply]
Exists @ `/usr/local/include/gme/gme.h`

EDIT: I could try symlinking `gme.h` to `/usr/local/include/` maybe...

EDIT 2: that worked along with `blargg_source.h`, but now it can't find `opusfile.h` despite it existing @ `/usr/include/opus/opusfile.h`

Hmm... yeah, it doesn't seem to like headers being in subdirectories. If I symlink them, it finds them.

**Reply by Wohlstand** [reply]
You need to install the libopusfile-dev

**Reply by Wohlstand** [reply]
About the GME, I should update the build, because there is a new public header was added for needs of the gme.h, and make sure the x/include/gme/ was added into the includes set

**Reply by Wohlstand** [reply]
I did the check and seeks its correct, the GME gets successfully detected and works fine :thinking: Speaking for the local build, do you trying to cross-compile the game from the desktop computer?

**Reply by ChristianSilvermoon** [reply]
Directly on the Pinephone over SSH

EDIT: this has been an adventure, but I think I'm gonna take a break for now and come back to it later... thanks so much for all the help, Wohlstand, you really are awesome ^^

**Reply by ChristianSilvermoon** [reply]
Okay, so I've come back to it, and I just cannot get SDL-Mixer-X to compile (using tag 2.2.0)

I've attached logs from my attempts.

I used `cmake -DCMAKE_INSTALL_PREFIX=/usr/local ..` from the build directory
and then `make -j4`

[sdl-mixer-x-2.2.0-cmake.log](https://github.com/Wohlstand/TheXTech/files/5939082/sdl-mixer-x-2.2.0-cmake.log)
[sdl-mixer-x-2.2.0-make.log](https://github.com/Wohlstand/TheXTech/files/5939083/sdl-mixer-x-2.2.0-make.log)

Am I using the wrong options or something?

EDIT: log was missing this bit (I should've redirected STDERR to STDOUT)
```
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/codecs/music_gme.c.o: in function `GME_setvolume':
music_gme.c:(.text+0x1c): undefined reference to `sqrt'
/usr/bin/ld: music_gme.c:(.text+0x34): undefined reference to `round'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/codecs/music_gme.c.o: in function `GME_jump_to_time':
music_gme.c:(.text+0xb0): undefined reference to `round'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_Timidity_addToPathList':
mixer_x_deprecated.c:(.text+0x0): undefined reference to `Mix_Timidity_addToPathList'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getBankNames':
mixer_x_deprecated.c:(.text+0x4): undefined reference to `Mix_ADLMIDI_getBankNames'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getBankID':
mixer_x_deprecated.c:(.text+0x10): undefined reference to `Mix_ADLMIDI_getBankID'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setBankID':
mixer_x_deprecated.c:(.text+0x14): undefined reference to `Mix_ADLMIDI_setBankID'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getTremolo':
mixer_x_deprecated.c:(.text+0x20): undefined reference to `Mix_ADLMIDI_getTremolo'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setTremolo':
mixer_x_deprecated.c:(.text+0x24): undefined reference to `Mix_ADLMIDI_setTremolo'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getVibrato':
mixer_x_deprecated.c:(.text+0x30): undefined reference to `Mix_ADLMIDI_getVibrato'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setVibrato':
mixer_x_deprecated.c:(.text+0x34): undefined reference to `Mix_ADLMIDI_setVibrato'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getScaleMod':
mixer_x_deprecated.c:(.text+0x40): undefined reference to `Mix_ADLMIDI_getScaleMod'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setScaleMod':
mixer_x_deprecated.c:(.text+0x44): undefined reference to `Mix_ADLMIDI_setScaleMod'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getAdLibMode':
mixer_x_deprecated.c:(.text+0x50): undefined reference to `Mix_ADLMIDI_getAdLibMode'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setAdLibMode':
mixer_x_deprecated.c:(.text+0x54): undefined reference to `Mix_ADLMIDI_setAdLibMode'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setSetDefaults':
mixer_x_deprecated.c:(.text+0x60): undefined reference to `Mix_ADLMIDI_setSetDefaults'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getLogarithmicVolumes':
mixer_x_deprecated.c:(.text+0x64): undefined reference to `Mix_ADLMIDI_getLogarithmicVolumes'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setLogarithmicVolumes':
mixer_x_deprecated.c:(.text+0x70): undefined reference to `Mix_ADLMIDI_setLogarithmicVolumes'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_getVolumeModel':
mixer_x_deprecated.c:(.text+0x74): undefined reference to `Mix_ADLMIDI_getVolumeModel'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_ADLMIDI_setVolumeModel':
mixer_x_deprecated.c:(.text+0x80): undefined reference to `Mix_ADLMIDI_setVolumeModel'
/usr/bin/ld: CMakeFiles/SDL2_mixer_ext.dir/src/mixer_x_deprecated.c.o: in function `MIX_OPNMIDI_setCustomBankFile':
mixer_x_deprecated.c:(.text+0x84): undefined reference to `Mix_OPNMIDI_setCustomBankFile'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/SDL2_mixer_ext.dir/build.make:354: lib/libSDL2_mixer_ext.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:97: CMakeFiles/SDL2_mixer_ext.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Built target SDL2_mixer_ext_Static
make: *** [Makefile:149: all] Error 2
```

**Reply by Wohlstand** [reply]
Please use the latest top  state, it has much better support of the stuff

**Reply by Wohlstand** [reply]
2.2.0 is too old

**Reply by ChristianSilvermoon** [reply]
Straight off the master branch? Should I attempt the same with all the dependencies?

EDIT: Attempting that with commit `b87b6f` off `master` got me....
[sdl-mixer-x-master-cmake.log](https://github.com/Wohlstand/TheXTech/files/5939102/sdl-mixer-x-master-cmake.log)
[sdl-mixer-x-master-make.log](https://github.com/Wohlstand/TheXTech/files/5939103/sdl-mixer-x-master-make.log)

**Reply by Wohlstand** [reply]
Dependencies can be older, that should be fine, but MixerX itself needs to be newer. Lemme review logs...

**Reply by Wohlstand** [reply]
I see, you have the older XMP than needed, please build the latest master of libXMP

**Reply by ChristianSilvermoon** [reply]
SUCCESS!
![20210207_07h46m24s_grim](https://user-images.githubusercontent.com/30162808/107150491-3fc76e00-691b-11eb-911c-bb9a4ba1757c.png)
![20210207_07h49m35s_grim](https://user-images.githubusercontent.com/30162808/107150496-45bd4f00-691b-11eb-9fc2-cdbb80f97bf0.png)
![20210207_07h50m01s_grim](https://user-images.githubusercontent.com/30162808/107150510-51107a80-691b-11eb-8bca-1dcd3984db6c.png)
![20210207_07h50m45s_grim](https://user-images.githubusercontent.com/30162808/107150502-4950d600-691b-11eb-8ab6-93f33e0d21c0.png)

Was able to play some using my USB Keyboard :D
Squeekboard (Phosh's Virtual Keyboard) sends the wrong key codes, I think, but I think that's a known bug that's been patched in later versions.

If some sort of on screen gamepad were possible, this could be perfect :D
Will have to test a phone gamepad sometime when I can get one

Thank you so, so very much for all the help, Wohlstand!

EDIT: FPS Averages 65. Except on Title at times, but a lot kinda happens on the Title Screen.

**Reply by Wohlstand** [reply]
Glad you built the thing 🦊
The Android target (at the latest master) has the touch-screen keyboard support, but it has some bugs I working on. Keyboards and USB / Bluetooth gamepads were supported by default. Enjoy the game 🦊

**Reply by Wohlstand** [reply]
Anyway, it's a good idea to allow enable the touchscreen controller optionally on any other non-Android builds (for example, the Desktop-like tablet also can be used too)

**Reply by ChristianSilvermoon** [reply]
Definitely would be a big help to have the touch controller as an option... maybe optionally enabled with `--touch-input=on` and `--touch-input=off` (could have a default specified at compile time, maybe? Or in a config?). Bluetooth is quite the battery vampire. So far my best experience playing on the Pinephone has been with my Left Joycon x3

**Comment by Wohlstand** [comment]
Since the libOPUS build got been fixed for AARCH64, now the common guide should work fine without such pain. :fox_face:
